feat: editable Base Point ComboBox added to the PointName Toolbar

Onetchou opened the pull request:
Hello everyone !

As discussed on the Seamly forum ([there](https://forum.seamly.io/t/construction-point-labelling/8925)), I implemented a Base Point ComboBox in the PointName Toolbar:
![image](https://github.com/FashionFreedom/Seamly2D/assets/83351185/e573f011-bdae-45d5-85a1-d79f51afe42a)

It allows the user to select another base name for every new point created, thus making the drafting of complexe blocks with a lot of construction points quicker and easier:
![image](https://github.com/FashionFreedom/Seamly2D/assets/83351185/aa677ea9-96a0-4466-9195-72c73ee54cc0)

The user can select it from the list ("default" uses the block base letter, as we are used to):
![image](https://github.com/FashionFreedom/Seamly2D/assets/83351185/5058dc0a-ac32-4f96-9054-ae6ff28dab07)

Or by entering directly the name into the box:
![image](https://github.com/FashionFreedom/Seamly2D/assets/83351185/419cb308-96f3-4a9b-8eeb-ac00a874cb1f)

By pressing enter, the user can save the custom name into the list to access it faster:
![image](https://github.com/FashionFreedom/Seamly2D/assets/83351185/2b2dcb3a-f587-4f92-8805-2b926d250890)

The combobox verifies the name with the same rules used in the tool dialogs (it uses the default name when an invalid name has been entered):
![image](https://github.com/FashionFreedom/Seamly2D/assets/83351185/1a21249f-fc3c-4dc0-b545-2eacc20be665)

The list in the combobox is updated according to the label language selected in the preferences:
![image](https://github.com/FashionFreedom/Seamly2D/assets/83351185/a27d3f37-4726-494a-827e-2bf7feab592f)
![image](https://github.com/FashionFreedom/Seamly2D/assets/83351185/6a44c33b-9659-4001-b43b-ea7be61d0241)

Since the modifications don't affect the dialogs, it works for every tool that creates new points, if I'm not mistaken.
As for the pen presets, it requires the user to select the base point name before selecting a tool. Maybe it could be useful, in a future release, to disable the pen preset comboboxes and the base point combobox when a tool has been selected to avoid confusion.

Cheers!

Onetchou committed:
feat: Base Point Combobox added to point name toolbar

Onetchou committed:
feat: editable base point combobox with data validation

Onetchou committed:
fix: base point combobox tooltip updated

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/FashionFreedom/Seamly2D?pullRequest=1064) <br/>All committers have signed the CLA.

ghost commented:
<details open>
<summary><b>👇 <a href="https://app.codesee.io/r/reviews?pr=1064&src=https%3A%2F%2Fgithub.com%2FFashionFreedom%2FSeamly2D">Click on the image</a> for a new way to code review</b>
</summary>

####

[![Review these changes using an interactive CodeSee Map](https://s3.us-east-2.amazonaws.com/maps.codesee.io/images/github/FashionFreedom/Seamly2D/1064/608e46a3/ab27e0a454746354366b2aa943c48838f3c0a296.svg)](https://app.codesee.io/r/reviews?pr=1064&src=https%3A%2F%2Fgithub.com%2FFashionFreedom%2FSeamly2D)

#### Legend
<img src="https://app.codesee.io/Legend_v3.png" alt="CodeSee Map legend" width="220" height="156">
</details>

DSCaskey commented:
Ok had a chance to build and test. In theory it works, but there are a few issues, and some unnecessary code. 

For starters you don't need to loop through the string list to add items - just add the whole string. 

```
    basePointComboBox->addItem(tr("Default"));

    QStringList alphabet = doc->GetCurrentAlphabet();
    for (int i = 0; i < alphabet.size(); i++)
    {
        basePointComboBox->addItem(alphabet.at(i));
    }
```
    
Can be simplified:

```
    basePointComboBox->addItem(tr("Default"));
    basePointComboBox->addItems(doc->GetCurrentAlphabet());
```

 Also the updating with the combobox is incorrect. These are unessesary as the Prefs are already emitting a QEvent::LanguageChange that mainwindow handles. 
 
mainwindow.cpp
```
     connect(qApp->Seamly2DSettings(), &VSettings::labelLanguageChanged, this, 
     &MainWindow::initBasePointComboBox);
```

vsettings.cpp
```
   emit labelLanguageChanged();
```

Language change events are handled in the override method: 

`virtual void changeEvent(QEvent* event) Q_DECL_OVERRIDE;`
    
```
    void MainWindow::changeEvent(QEvent *event)
{
    if (event->type() == QEvent::LanguageChange)
```

Only need to call  initBasePointComboBox() in  changeEvent():

```    
 UpdateWindowTitle();
 initPenToolBar();
 initBasePointComboBox();
```
These perform no function:  setter is called but getter is never called. 

```
    void                           setDefaultBasePoint(QString basePoint);
    QString                        getDefaultBasePoint() const;

mainwindow()
    //we keep basePoint empty if default is selected
    doc->setDefaultBasePoint(basePoint);
```
That may be my bad as you were probably just following the penChanged() handler. I think my idea was to be able to pass the whole Pen() and not have to pass the color, linetype & weight. 

Also, since there are changes to the UI, we need to run  Lupdate to update the translation ts files.

Onetchou commented:
> For starters you don't need to loop through the string list to add items - just add the whole string.

Ok, I'll update that.

> Not sure why... but the combobox is not adding the items based on the current language? I set the language to Russian and still got this:

I think this is because you are changing the GUI language and not the label language. In the current seamly version there are two language parameters, one for the interface and the other for the labels:
![image](https://github.com/FashionFreedom/Seamly2D/assets/83351185/5dfddc28-c368-4a12-8bd1-80dcad5da0e1)

The label language allows to have the points in one language and the interface in another:
![image](https://github.com/FashionFreedom/Seamly2D/assets/83351185/8a7a2a17-a875-4808-bdf4-b1d7b41c40e6)

Or the other way around:
![image](https://github.com/FashionFreedom/Seamly2D/assets/83351185/e907946c-aa91-4cdf-a494-0acabd8ad3bc)


I don't know why this distinction has been made but with the current version of Seamly if you only change the GUI language in Russian, the labels are still created using the American alphabet, I tried to follow the same logic, this is why the combobox stays in english when only the GUI language is modified.

Note that the "Default" would be translated in Russian if you change the GUI language, though.

> Also the updating with the combobox is incorrect. These are unessesary as the Prefs are already emitting a QEvent::LanguageChange that mainwindow handles.

This is the GUI language, not the label language.
Using QEvent::LanguageChange would make the points be translated with the GUI language, thus removing (part of?) the utility of the label language, don't you think?

Onetchou commented:
> These perform no function: setter is called but getter is never called.

I thought that coding a setter without a getter was strange so I added the getter too with the idea that it may find an utility in the future.
I'll remove it if you prefer.

Onetchou committed:
refactor: unused getDefaultBasePoint method removed and code simplification

Onetchou commented:
It seems that the label language has no other utility than the one detailed above:
![image](https://github.com/FashionFreedom/Seamly2D/assets/83351185/11b853c5-c432-472c-9e4b-16f4fa97f575)

If you agree that it is not useful I can remove this setting so there's only one language option for everything, thus linking everything to the GUI language (that could thus be renamed simply "language").

Onetchou committed:
feat: labels are now using the GUI language instead of the label language

Onetchou committed:
feat: label language combobox has been removed

Onetchou commented:
I removed the label language so there is only one 'Language" combobox in the preferences that controls everything:
![image](https://github.com/FashionFreedom/Seamly2D/assets/83351185/3d6f9d98-8c6f-4427-9e64-883cc7b6a3c0)

Tell me if I have to revert the changes if the label language was useful for some reasons :blush:

DSCaskey commented:
>I removed the label language so there is only one 'Language" combobox in the preferences that controls everything:

NOOOO.  There's a reason why there's both.  It allows a users to use the UI in one language and print patterns in another.

DSCaskey commented:
>I think this is because you are changing the GUI language and not the label language. In the current seamly version there are two language parameters, one for the interface and the other for the labels:

Yes... I realized that... that's why I edited the comments.

Onetchou commented:
> > I removed the label language so there is only one 'Language" combobox in the preferences that controls everything:
> 
> NOOOO. There's a reason why there's both. It allows a users to use the UI in one language and print patterns in another.

I'm sorry but I don't get it... How is this label language working in the current version of Seamly? I mean from the user's perspective.

The name of the points are not printed on the patterns, so they are used by the designer, so there's no reason to use a different language than the GUI, right?
And the translatable things on the pattern are the labels, which are not translated by the "label language"? Would I have missed a connection between the label language and the labels?

DSCaskey commented:
The label language is determined by the locale()... of which there are only there are only 7 - they are NOT translated.  The GUI language - of which there are 20 - are translated. 

What do you think would happen if you start translating point names? Formulas would break. 

Please. Leave the label languages as is

Onetchou commented:
I think there is a big misunderstanding between us there, partly because there's a confusion between labels (piece / pattern labels) and point names (that are affected by the label language).
I obviously never though of translating the point names.

Just to recap, correct me if I'm wrong:
* The feature I added affects the point names. These points are not translated, once they have been added, they keep the same name.
* With the current seamly version, the labels of the points generated by default by the tools are taken in the alphabet defined in the "label language" preference.
* In order to keep the same logic, my combobox was filled with letters based on the corresponding "label language" alphabet.

Modulo the small refactoring to remove the unused getter and to simplify the creation of the combobox, this was the initial commit I suggested : n°[1eca1dd](https://github.com/FashionFreedom/Seamly2D/pull/1064/commits/1eca1dd7f45e8f9432dca4aa9f11fb6989f0535c)

* Then you told me to use the  QEvent::LanguageChange event, which is emitted when the GUI language is changed:

> Also the updating with the combobox is incorrect. These are unessesary as the Prefs are already emitting a QEvent::LanguageChange that mainwindow handles.

> Only need to call initBasePointComboBox() in changeEvent():

It means that it should be updated with the GUI language, and thus it makes no sense to continue using  the "label language" to select the proper alphabet as it was before. I will not update the point language with the value in the "label language" combobox only when the other "GUI combobox" is updated...  I thus implemented the changes in the following commit:
n°[13843d6](https://github.com/FashionFreedom/Seamly2D/pull/1064/commits/13843d62f68cc1b50542798410797891d0c955eb)

* Since, if I'm not mistaken, the label language was only used to select the alphabet used for the points label, it has no more utility now that the GUI language is used to do so:
> It seems that the label language has no other utility than the one detailed above
![image](https://github.com/FashionFreedom/Seamly2D/assets/83351185/7726e84d-37db-493b-a31c-d16099f02a34)


* I thus removed the combobox in this commit:
[ab27e0a](https://github.com/FashionFreedom/Seamly2D/pull/1064/commits/ab27e0a454746354366b2aa943c48838f3c0a296)

What I don't undersand is:
* > NOOOO. There's a reason why there's both. It allows a users to use the UI in one language and print patterns in another

I don't get how using point names in a certain language is useful to print the patterns in another language since the point names are not printed on the pattern. 
What I meant by evoking the translation of labels (pattern /piece labels) is that the information on the printed patterns which are language dependant are the pattern/piece labels, and, unless I'm wrong, I didn't find any connection in the code between the "label language" preference and a feature allowing to have multilanguage pattern/piece labels.

Since the name of the points are only used by the designer of the pattern, It makes sense that he uses the same language for the points and for the GUI, as you suggested by telling me to use the QEvent::LanguageChange event. I understand the fact that there are fewer languages for the points than for the GUI, this is why in my last commit the alphabet corresponding to the GUI language is chosen if it is available among the list used for point names, if not it chooses the "American" alphabet.

DSCaskey commented:
>I think there is a big misunderstanding between us there, partly because there's a confusion between labels (piece / pattern labels) and point names (that are affected by the label language).
I obviously never though of translating the point names.

Again... this is a result of the original dev's native language is not English. Some, but not all the code I've renamed to PointName vs Label... there were some really quirky names like labelPointLabel vs pointName_Label.  To him everything was a label. Point  name (text) were labels, QLabels were labels, pattern piece labels were labels... and if you look at Valentina... symbols are labels. 

![image](https://github.com/FashionFreedom/Seamly2D/assets/31944718/fbdd938f-b9ae-4182-a292-9566c5bc1a5a)

![image](https://github.com/FashionFreedom/Seamly2D/assets/31944718/8592701b-64fc-4256-9bac-3947fac612fb)

![image](https://github.com/FashionFreedom/Seamly2D/assets/31944718/4118cc89-e71a-4b5c-b8f3-12fe3b010883)

All very confusing. 

> The feature I added affects the point names. These points are not translated, once they have been added, they keep the same name.

Correct.

> With the current seamly version, the labels of the points generated by default by the tools are The feature I added affects the point names. These points are not translated, once they have been added, they keep the same name.taken in the alphabet defined in the "label language" preference.

Correct

> In order to keep the same logic, my combobox was filled with letters based on the corresponding "label language" alphabet.

Correct

>Since the name of the points are only used by the designer of the pattern, It makes sense that he uses the same language for the points and for the GUI, as you suggested by telling me to use the QEvent::LanguageChange event. I understand the fact that there are fewer languages for the points than for the GUI, this is why in my last commit the alphabet corresponding to the GUI language is chosen if it is available among the list used for point names, if not it chooses the "American" alphabet.

I get it, but by removing the (Point name) Label language, you've removed that option that some users may use. In other words you've changed the behavior of the app. I'm not opposed to removing the Point name language IF there's a consensus of the users on the forums. I never want to do anything that takes something away from a user. 

>>I don't get how using point names in a certain language is useful to print the patterns in another language since the point names are not printed on the pattern.

That all being said... I actually see a use for a Label  and Symbol language, where the user can actually work in a GUI language, and yet print patterns with symbols, text and labels in another. So in that case we would come full circle and truely have need for a "Label" or print language... without having to reset the GUI prefs to do so or as it is now type the labels in another language. Problem here is how to deal with the placeholder data that may translated from various UI sources... for example a rotation of None in the combobox in Piece Dialog->Properties.

Onetchou committed:
revert: basepoint labels use label language

Revert the following commits :
"feat: label language combobox has been removed" n°ab27e0a454746354366b2aa943c48838f3c0a296
"feat: labels are now using the GUI language instead of the label language" n° 13843d62f68cc1b50542798410797891d0c955eb

Onetchou committed:
feat: renaming the "label combobox" into a "point name alphabet" combobox

Onetchou commented:
> everything was a label. Point name (text) were labels, QLabels were labels, pattern piece labels were labels... and if you look at Valentina... symbols are labels.`

Aoutch 😂

> I'm not opposed to removing the Point name language IF there's a consensus of the users on the forums.
> I actually see a use for a Label and Symbol language, where the user can actually work in a GUI language, and yet print patterns with symbols, text and labels in another. So in that case we would come full circle and truely have need for a "Label" or print language...

I think we're on the same page now 😊

What I propose:
* I have reverted my last two commits, so the new basepoint combobox is using the "label language".
* I renamed the "label language" combobox into a "point name alphabet" combobox, since that's currently it's only use. It avoids confusion between pattern/piece labels and point name labels. I think using the word "alphabet" instead of "language" removes the expectation from the user to see his point names translated if he's using words for point names:
![image](https://github.com/FashionFreedom/Seamly2D/assets/83351185/215e49e5-251a-444e-8bc1-5ca99cd3e7de)
* I open a thread on the forum to discuss about these translation preferences, including the potential fusion of the GUI and point name language, at least until a true way to translate patterns is implemented (in that case the best option would probably be to keep the label language in the code but only hide the combobox from the menu, this way it would be ready when a proper pattern translation feature is added to Seamly).

DSCaskey commented:
>I renamed the "label language" combobox into a "point name alphabet" combobox

 like the idea of Point name... although to be consistent I would use Point name text: As point names can be more than just a letter. 
 
![Screenshot 2024-02-26 115944](https://github.com/FashionFreedom/Seamly2D/assets/31944718/41c38ebc-318d-4a58-a939-18a3672f7c1b)

>I open a thread on the forum to discuss about these translation preferences, including the potential fusion of the GUI and point name language, at least until a true way to translate patterns is implemented (in that case the best option would probably be to keep the label language in the code but only hide the combobox from the menu, this way it would be ready when a proper pattern translation feature is added to Seamly).

I would say if there is no disagreement to remove the "label" (point name) language to use the GUI language for point names as in the manner you used. At some point I would add the option for a "Label", or if you will an "Export" language, and add the dropdown in the Export dialog.  Also like the format type, I would save the language in the Prefs so that the "Last used" language is the default. This would then allow a user to select what language to Export the pattern in. That being said there would be a bunch of underlying code to change the language for the Labels and future Text and Symbol tools.

Onetchou committed:
feat: "point name alphabet" combobox renamed "point name text" combobox

Onetchou commented:
> like the idea of Point name... although to be consistent I would use Point name text: As point names can be more than just a letter.

Got it, it has been renamed in this last commit:
![image](https://github.com/FashionFreedom/Seamly2D/assets/83351185/283fa622-b15b-4098-8806-1eab0cce6a72)

> At some point I would add the option for a "Label", or if you will an "Export" language, and add the dropdown in the Export dialog. Also like the format type, I would save the language in the Prefs so that the "Last used" language is the default. This would then allow a user to select what language to Export the pattern in. That being said there would be a bunch of underlying code to change the language for the Labels and future Text and Symbol tools.

An "export language", or maybe rather a "pattern language" combobox in the interface of the piece editor would be useful too in order to have a preview, or even in order to allow the user to write different (piece / pattern) label texts for each language if he wants to translate them.

DSCaskey commented:
@Onetchou 

In our weekly meeting I discussed the label / point name language issue.  I knew there was a reason why it's there, and whyI didn't want to just remove it. Let's say I work in Ukrainian for the GUI, but I need to share the pattern with others that need the pattern in English. All the point name base letters in the formulas are going to be in Ukrainian , not English.  If the pattern is to be shared  9 times out of 10 you're going to want the point names in English.  Of course this is only applicable to the base letter, as anything else you type in a Name field is going to be in the GUI language.

Onetchou commented:
> @Onetchou
> 
> In our weekly meeting I discussed the label / point name language issue. I knew there was a reason why it's there, and whyI didn't want to just remove it. Let's say I work in Ukrainian for the GUI, but I need to share the pattern with others that need the pattern in English. All the point name base letters in the formulas are going to be in Ukrainian , not English. If the pattern is to be shared 9 times out of 10 you're going to want the point names in English. Of course this is only applicable to the base letter, as anything else you type in a Name field is going to be in the GUI language.

Indeed! I hadn't thought of that case. Thanks for keeping me informed 😊

DSCaskey commented:
>Indeed! I hadn't thought of that case. Thanks for keeping me informed 

I knew I recalled reading some discussion on the topic in the forums before I joined... just couldn't exactly remember what it was. At any rate,  since there are changes to the UI do you know how to run an Lupdate? If not I can do it as it only takes like 10 secs to run from Creator. Once we have committed the updated ts files I can approve & run, and if everything passes I'll merge.

Onetchou committed:
fix: .ts files updated

Onetchou commented:
> At any rate, since there are changes to the UI do you know how to run an Lupdate? If not I can do it as it only takes like 10 secs to run from Creator. Once we have committed the updated ts files I can approve & run, and if everything passes I'll merge.

Here they are 😊
